nucypher/utilities/metrics.py

```diff
@@ -1,23 +1,74 @@
 try:
-    from prometheus_client import Gauge, Enum, Counter, Info
+    from prometheus_client import Gauge, Enum, Counter, Info, Histogram, Summary
 except ImportError:
     raise ImportError('prometheus_client is not installed - Install it and try again.')
 from twisted.internet import reactor, task
 
 import nucypher
-from nucypher.blockchain.eth.agents import ContractAgency, StakingEscrowAgent
+from nucypher.blockchain.eth.agents import ContractAgency, StakingEscrowAgent, WorkLockAgent, PolicyManagerAgent
+from nucypher.blockchain.eth.actors import NucypherTokenActor
+from typing import List
+from nucypher.blockchain.eth.interfaces import BlockchainInterfaceFactory
 
-# Metrics
-known_nodes_guage = Gauge('known_nodes', 'Number of currently known nodes')
-work_orders_guage = Gauge('work_orders', 'Number of accepted work orders')
-missing_confirmation_guage = Gauge('missing_confirmations', 'Currently missed confirmations')
-learning_status = Enum('node_discovery', 'Learning loop status', states=['starting', 'running', 'stopped'])
-requests_counter = Counter('http_failures', 'HTTP Failures', ['method', 'endpoint'])
-host_info = Info('host_info', 'Description of info')
-active_stake_gauge = Gauge('active_stake', 'Active stake')
 
+class BaseEventMetricsCollector:
 
-def collect_prometheus_metrics(ursula):
+    def __init__(self, staker_address, worker_address, contract_agent, event_name, argument_filters, metrics):
+        self.event_filter = contract_agent.contract.events[event_name].createFilter(fromBlock='latest',
+                                                                                    argument_filters=argument_filters)
+        self.metrics = metrics
+        self.event_name = event_name
+        self.staker_address = staker_address
+        self.worker_address = worker_address
+        self.contract_agent = contract_agent
+
+    def collect(self, node_metrics):
+        events = self.event_filter.get_new_entries()
+        for event in events:
+            self.event_occurred(event, node_metrics)
+
+    def event_occurred(self, event, node_metrics):
+        for arg in self.metrics.keys():
+            if arg == "block_number":
+                self.metrics["block_number"].set(event["blockNumber"])
+                continue
+            self.metrics[arg].set(event['args'][arg])
+
+
+class ReStakeEventMetricsCollector(BaseEventMetricsCollector):
+    def __init__(self, staker_address, worker_address, contract_agent, event_name, argument_filters, metrics):
+        super().__init__(staker_address, worker_address, contract_agent, event_name, argument_filters, metrics)
+        self.metrics["reStake"].set(self.contract_agent.is_restaking(staker_address))
```

Comment:
Is there a reason that the `reStake` metric (and some others) are only checked and set on initialization and not during `event_occurred`? I'm sure there is a reason... 😅

Reply:
Yeah, there is :)
This metrics will be showed immediately after the start, so we don't need to wait for the events to show them in the Prometheus.

Reply:
But they will still be checked in event_occurred() for further updates

Reply:
Right! 👍